_posts/2016-06-08-component-relative-paths-in-angular-2.md

```diff
@@ -0,0 +1,305 @@
+
+Consider a simple *Header* component... which we actually use in our Angular 2 Master Class training.
+
+**header.component.ts**
+{% highlight ts %}
+{% raw %}
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'contacts-header',
+  template: `
+    <nav class="navbar-fixed">
+      <div class="nav-wrapper">
+        <span class="brand-logo center">Contacts</span>
+      </div>
+    </nav>
+  `,
+  styles: ['.navbar-fixed { position:fixed; }']
+})
+export class HeaderComponent implements OnInit {
+  constructor() {}
+  ngOnInit() {}
+}
+{% endraw %}
+{% endhighlight %}
+
+Using this component is super easy and since there are no external file dependencies and everything is defined inline then you should *never* see a 404 error [for this component] in the DevTools console.
+
+## Components with External Assets
+
+Another options let's us load html and styles from external files by using the URLs in the metadata configuration block. This is a common practice to split a component's code, HTML, and CSS into three separate files in the same directory:
+
+*  header.component.ts
+*  header.component.html
+*  header.component.css
+
+This practice of using external files is especially important when your HTML or CSS is non-trivial. Using external files keeps your *.ts logic files much cleaner and easier to maintain.
+
+**header.component.ts**
+{% highlight ts %}
+{% raw %}
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector   : 'contacts-header',
+  templateUrl: 'header.component.html',
+  stylesUrl  : 'header.component.css'
+})
+export class HeaderComponent implements OnInit {
+  constructor() {}
+  ngOnInit() {}
+}
+{% endraw %}
+{% endhighlight %}
+
+**header.component.css**
+{% highlight css %}
+{% raw %}
+.navbar-fixed {
+  position:fixed;
+}
+{% endraw %}
+{% endhighlight %}
+
+**header.component.html**
+{% highlight html %}
+{% raw %}
+<nav class="navbar-fixed">
+  <div class="nav-wrapper">
+    <span class="brand-logo center">Contacts</span>
+  </div>
+</nav>
+{% endraw %}
+{% endhighlight %}
+
+Now this is where it gets tricky!
+
+The URL required is relative to the application root which is usually the location of the `index.html` web page that hosts the application. So the above component must be stored in the application root. Wow, this does not scale well.
+
+*  What if we have many components all at the root level ? ( OMG! )
+*  What if my components are organized in distinct packages ?
+*  What if we want to organize our components by feature (Angular Style Guide - Best Practice)?
+
+To explore the issue, let's consider the scenario where our details component (and files) are in the `src/app/header` package. The urls used above would cause the loader to fail and the developer would see the following 404 error in the developers console:
+
+![component_url_404](https://cloud.githubusercontent.com/assets/210413/15878482/eca5cba2-2ce0-11e6-8fb4-78868bec2644.png)
```

Comment:
I try to use a little images as possible when it comes to code snippets and errors as they don't scare very well. 

E.g. if the error message will change, this image is outdate and we need to create new screenshots and everything. Text however, we can just change.

I'd propose to just add this:

```
EXCEPTION: Failed to load header.component.html
```

(just the subject of the exception without the stack trace and everything)